src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -1830,12 +1830,44 @@ private static BoundExpression SkipReferenceConversions(BoundExpression possibly
             return possiblyConversion;
         }
 
-        // https://github.com/dotnet/roslyn/issues/30140: Track nullable state from ??=.
         public override BoundNode VisitNullCoalescingAssignmentOperator(BoundNullCoalescingAssignmentOperator node)
         {
-            var result = base.VisitNullCoalescingAssignmentOperator(node);
-            SetResult(node);
-            return result;
+
+            BoundExpression leftOperand = node.LeftOperand;
+            BoundExpression rightOperand = node.RightOperand;
+            int leftSlot = MakeSlot(leftOperand);
+
+            // The assignment to the left below needs the declared type from VisitLvalue, but the hidden
+            // unnecessary check diagnostic needs the current adjusted type of the slot
+            VisitLvalue(leftOperand);
+            TypeSymbolWithAnnotations leftResult = _resultType;
+            TypeSymbolWithAnnotations currentLeftResult = GetAdjustedResult(leftResult, leftSlot);
+
+            if (currentLeftResult.ValueCanBeNull() == false)
+            {
+                ReportNonSafetyDiagnostic(ErrorCode.HDN_ExpressionIsProbablyNeverNull, leftOperand.Syntax);
+            }
+
+            var leftState = this.State.Clone();
+
+            if (leftSlot > 0)
+            {
+                leftState[leftSlot] = NullableAnnotation.NotNullable;
```

Comment:
Yes, otherwise nothing will tell `Join` that the slot is non-nullable in the left path, and when it attempts to join the slot from the left and right sides, it will not correctly conclude that the slot is non-nullable.

---
In reply to: [248489590](https://github.com/dotnet/roslyn/pull/32225#discussion_r248489590) [](ancestors = 248489590)

Reply:
> otherwise nothing will tell Join that the slot is non-nullable in the left path, and when it attempts to join the slot from the left and right sides, it will not correctly conclude that the slot is non-nullable.

Why do we care what value is going to be in the slot after the Join? It looks like we are overriding that value immediately after the Join.

---
In reply to: [249132309](https://github.com/dotnet/roslyn/pull/32225#discussion_r249132309) [](ancestors = 249132309,248489590)